strawberryfields/engine.py

```diff
@@ -609,15 +610,63 @@ def run_async(self, program: Program, *, compile_options=None, **kwargs) -> Job:
         """
         # get the specific chip to submit the program to
         compile_options = compile_options or {}
+        recompile = compile_options.pop("recompile", False)
         kwargs.update(self._backend_options)
 
         device = self.device_spec
 
         compiler_name = compile_options.get("compiler", device.default_compiler)
-        msg = f"Compiling program for device {device.target} using compiler {compiler_name}."
-        self.log.info(msg)
 
-        program = program.compile(device=device, **compile_options)
+        if program.compile_info is None:
+            # program is uncompiled
+            msg = f"Compiling program for device {device.target} using compiler {compiler_name}."
+            self.log.info(msg)
+            program = program.compile(device=device, **compile_options)
+
+        elif (
+            program.compile_info[0].target != device.target
+            or program.compile_info[0]._spec != device._spec
+        ):
+            # program was compiled for a different device
+            if not recompile:
+                raise ValueError(
+                    "Cannot use program compiled for "
+                    f"{program._compile_info[0].target} for target "
+                    '{self.target}. Set the "recompile" entry in the '
+                    "compile_options dictionary to true to compile with "
+                    "{compiler_name}."
+                )
+
+            msg = f"Recompiling program for device {device.target} using compiler {compiler_name}."
+            self.log.info(msg)
+            program = program.compile(device=device, **compile_options)
+
+        elif program.compile_info[1] != compiler_name:
+            # program was compiled for the device, but using a different compiler
+
+            if compile_options is None or not compile_options:
+                msg = (
+                    f"No compile options specified, compiling program for "
+                    f"device {device.target} using compiler Xstrict."
+                )
+                self.log.info(msg)
+
+                program = program.compile(device=device, compiler="Xstrict")
+            else:
+                # User has explicitly provided compile_options to the engine
+                msg = (
+                    f"Compiling program for device {device.target}"
+                    f"using the specified compiler options."
+                )
+                self.log.info(msg)
+
+                program = program.compile(device=device, **compile_options)
+        else:
+            msg = (
+                f"No compilation, compiled program compatible with device "
+                f"{device.target} and compiler {compiler_name}."
+            )
+            self.log.info(msg)
```

Comment:
Does this `else` correspond to 'Program was compiled for the device using the default device compiler'? Perhaps this should be combined with the previous `elif`, so that

* The program is still validated against `Xstrict`
* The program is recompiled if `recompile=True` and `compile_options` are provided (as far as I can tell, this won't happen here; a program already compiled against the default compiler won't be recompiled even if the user provides `compile_options`).

Reply:
An addition to this: wouldn't it make sense to _always_ recompile if `recompile=True`? In that case this could be outside/above of most of this logic, and simply jump over it if `recompile=True`.

Reply:
Thanks for the comments! @thisac that's a fair point, one reason to keep it is as is to be able to log messages that are more specific to the case

Reply:
Reiterated on this point @thisac, it made me rethink how these branches of the if statement could align together. Thanks again :)